### DISCUSSION TITLE
Do not fail when there is no internet connection

### DISCUSSION THREAD

**Discussion by danielo515** [twpayne/chezmoi#2622] [Ideas]
Hello.
I love chezmoi, and I use it a lot, but it always drives me nuts when I just want to apply a file and I get a long stack-trace because network connectivity.
Will it be possible to detect network connectivity and ignore all apply operations that require a network? Or at least perform this checks only when they make more sense? For example, I use a lot the apply to a single file, so it does not make much sense to do any network request in that case, because it is a local file.

Thanks!

**Comment by twpayne** [comment]
Could you give an example of the stack trace you get?

You can pass the `--refresh-externals=false` option to stop chezmoi from trying to update externals.

**Reply by danielo515** [reply]
This is one I also get when I do some commands. I want to clarify that I call chezmoi from a NeoVim autocmd, hence the stack-trace is mixed with lua stuff

```
Error detected while processing BufWritePost Autocommands for "danielo_nvim/*.hx":
Error executing lua callback: .../site/pack/packer/start/plenary.nvim/lua/plenary/job.lua:405: Failed to spawn process: {
  _additional_on_exit_callbacks = {},
  _raw_cwd = 'chezmoi: /Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml: template: /Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml:35:65: executing "/Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml" at <gitHubLatestRelease "zsh-users/zsh-completions">: error calling gitHubLatestRelease: Get "https://api.github.com/repos/zsh-users/zsh-completions/releases/latest": dial tcp 140.82.121.5:443: connect: connection refused/../danielo_nvim',
  
stack traceback:
	.../site/pack/packer/start/plenary.nvim/lua/plenary/job.lua:405: in function '_execute'
	.../site/pack/packer/start/plenary.nvim/lua/plenary/job.lua:448: in function 'start'
	/Users/danielo/.config/lvim/lua/user/autocommands.lua:37: in function </Users/danielo/.config/lvim/lua/user/autocommands.lua:30>
stack traceback:
	[C]: in function 'error'
	.../site/pack/packer/start/plenary.nvim/lua/plenary/job.lua:405: in function '_execute'
	.../site/pack/packer/start/plenary.nvim/lua/plenary/job.lua:448: in function 'start'
	/Users/danielo/.config/lvim/lua/user/autocommands.lua:37: in function </Users/danielo/.config/lvim/lua/user/autocommands.lua:30>
17:54:13 [DEBUG] lvim: 'Using chezmoi: /Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml: template: /Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml:35:65: executing "/Users/danielo/.local/share/chezmoi/chezmoi/.chezmoiexternal.toml" at <gitHubLatestRelease "zsh-users/zsh-completions">: error calling gitHubLatestRelease: Get "https://api.github.com/repos/zsh-users/zsh-completions/releases/latest": dial tcp 140.82.121.5:443: connect: connection refused as chezmoi path' file="chezmoi.lua", line=9, title="Danielo"
"danielo_nvim/src/vim/Vim.hx" line 74 of 79 --93%-- col 40-43
"danielo_nvim/src/vim/Vim.hx" [+] line 74 of 78 --94%-- col 40-43
"danielo_nvim/src/vim/Vim.hx" 78L, 1971B [w]
17:54:55 [DEBUG] lvim: "Using /Users/danielo/.local/share/chezmoi/chezmoi as chezmoi path" file="chezmoi.lua", line=9, title="Danielo"
```

**Comment by twpayne** [comment]
Ah, I see you're using the `gitHubLatestRelease` template function in your `.chezmoiexternal.toml`. This requires network access, and there's not much that chezmoi can do if you don't have network access.

One work-around would be to make the call to `gitHubLatestRelease` in your config file template, store it in a variable in the `data` section, and then use the variable in your `.chezmoiexternal.toml` file. This will mean that network access is only needed when you run `chezmoi init`, not when you run `chezmoi apply`.

**Reply by danielo515** [reply]
Do I? Hahaha, can I post a link to my configuration and you point me where am I doing it wrong?

**Reply by twpayne** [reply]
Sure.

**Reply by danielo515** [reply]
Here is my repo: https://github.com/danielo515/dotfiles

**Reply by twpayne** [reply]
You're using the `gitHubLatestRelease` template function in your `.chezmoiexternal.toml` file [here](https://github.com/danielo515/dotfiles/blob/e6bdac30fa9ca752799b39fed6cbb56ae12c9bb5/chezmoi/.chezmoiexternal.toml#L35) and [here](https://github.com/danielo515/dotfiles/blob/e6bdac30fa9ca752799b39fed6cbb56ae12c9bb5/chezmoi/.chezmoiexternal.toml#L42).

The easiest fix is to hardcode the version number instead, as you do for your other externals.

**Comment by danielo515** [comment]
Oh, I see.
There is no way to make `gitHubLatestRelease` use a cached version, or skip
it completely if there is no internet?

On Wed, Dec 14, 2022 at 11:58 AM Tom Payne ***@***.***> wrote:

> You're using the gitHubLatestRelease template function in your
> .chezmoiexternal.toml file here
> <https://github.com/danielo515/dotfiles/blob/e6bdac30fa9ca752799b39fed6cbb56ae12c9bb5/chezmoi/.chezmoiexternal.toml#L35>
> and here
> <https://github.com/danielo515/dotfiles/blob/e6bdac30fa9ca752799b39fed6cbb56ae12c9bb5/chezmoi/.chezmoiexternal.toml#L42>
> .
>
> The easiest fix is to hardcode the version number instead, as you do for
> your other externals.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/twpayne/chezmoi/discussions/2622#discussioncomment-4398891>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWN6MPEVJFXSRECJ5HTWNGR47ANCNFSM6AAAAAAS3CRII4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com

**Reply by twpayne** [reply]
How would you define "no internet"?

**Comment by danielo515** [comment]
That's fair.
Will it be possible to catch the exception and just display a warning?
After all, it is just that you can not apply those "updates". I don't see
how that error preventing the rest of your dotfiles being applied is
desirable for anyone.

On Wed, Dec 21, 2022 at 11:09 PM Tom Payne ***@***.***> wrote:

> How would you define "no internet"?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/twpayne/chezmoi/discussions/2622#discussioncomment-4471233>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWIDWH47NBLMBXDWI3LWON5YRANCNFSM6AAAAAAS3CRII4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com

**Reply by twpayne** [reply]
In this thread I have given you instructions on how to make `chezmoi apply` work without an internet connection. Please follow these instructions.
